index.js

```diff
@@ -1,107 +1,80 @@
-var assert = require("assert");
+const { assertFunction, assertTitle } = require("./lib/util");
+const {
+  queueHandler,
+  taskQueue,
+  taskType,
+  taskAddedEventName
+} = require("./lib/queue");
+const { metaData } = require("./lib/execute");
 
-var d = null;
-var doPrint = false;
-var its = [];
-var beforeEachs = [];
-var befores = [];
-var afters = [];
-var afterEachs = [];
-var printFn = console.log.bind(console);
-
-function it(title, fn) {
-  assert(typeof title === "string", "title required");
-  assert(typeof fn === "function", "missing function");
-
-  its.push(function() {
-    if (fn.length) return unsupported("async/callback testing")();
-
-    beforeEachs.forEach(run);
-    try {
-      fn();
-      log(title, null);
-    } catch (err) {
-      log(title, err.message);
-    }
-    afterEachs.forEach(run);
-  });
-
-  // To support running without outer describe block
-  if (!d) {
-    its.forEach(run);
-    its = [];
-  }
+function after(fn) {
+  after.caller[metaData].afterHook = fn;
 }
 
-function describe(title, fn) {
-  assert(title, "title required");
-  assert(typeof fn === "function");
-
-  d = title;
-  fn();
-  befores.forEach(run);
-  its.forEach(run);
-  afters.forEach(run);
-
-  // reset
-  its = [];
-  befores = [];
-  afters = [];
-  afterEachs = [];
-  beforeEachs = [];
-  d = null;
+function afterEach(fn) {
+  assertFunction(fn);
+  afterEach.caller[metaData].afterEachHookCollection.push(fn);
 }
 
 function before(fn) {
-  befores.push(fn);
-}
-
-function after(fn) {
-  afters.push(fn);
+  assertFunction(fn);
+  before.caller[metaData].beforeHook = fn;
 }
 
 function beforeEach(fn) {
-  beforeEachs.push(fn);
+  assertFunction(fn);
+  beforeEach.caller[metaData].beforeEachHookCollection.push(fn);
 }
 
-function afterEach(fn) {
-  afterEachs.push(fn);
-}
+function describe(title, fn) {
+  assertTitle(title);
+  assertFunction(fn);
 
-function log(title, err) {
-  var status = err ? `❌. ${err}` : "✔️";
-  if (doPrint) {
-    printFn((d ? `${d}: ` : "") + `${title}: ${status}`);
-  }
-}
-function run(fn) {
-  fn();
-}
-function unsupported(title) {
-  return function() {
-    console.error("This operation is unsupported: " + title);
+  fn[metaData] = {
+    title,
+    afterHook: null,
+    afterEachHookCollection: [],
+    beforeHook: null,
+    beforeEachHookCollection: [],
+    itCollection: []
   };
+
+  taskQueue.push({
+    type: taskType.describe,
+    fn
+  });
+
+  queueHandler.emit(taskAddedEventName);
 }
 
-function install() {
-  doPrint = true;
-  global.describe = describe;
-  global.it = it;
-  global.before = before;
-  global.after = after;
-  global.beforeEach = beforeEach;
-  global.afterEach = afterEach;
+function it(title, fn) {
+  assertTitle(title);
+  assertFunction(fn);
+
+  const caller = it.caller ? it.caller[metaData] : undefined;
```

Comment:
I thought of keeping the each `it`s and `eachHooks` inside the function. I set the `metaData` attribute in the caller function. https://github.com/fatso83/mini-mocha/blob/5ebbb3e823a55206a6b28d0a8c3cd784c7960bf5/index.js#L33
so each hook will update the `metaData` attribute of its caller. Do you have any other suggestion?

Reply:
The code was _fine_. I was just suggesting to add some comments here and there. It's a bit sparse on the technical documentation, that's all.

Reply:
Thanks for the explanation. I think that should have been a comment in the code. It would have made it much easier :-)

No, I don't have a suggestion for improvement as it is, but I usually refrain from using unsupported props like that. Instead I try to explicitly pass what I need in, in some way, to avoid it. Not totally sure how that would work here, but it's probably doable.